### ISSUE TITLE
Secure connections without CA and client certificates. 

### ISSUE THREAD

**Issue by ivaylo-valkov-wrk** [alanxz/SimpleAmqpClient#178]
It should be possible to create SSL connections without providing CA and client certificates and keys.

This is a real use case scenario. The current Channel (master branch, Channel.cpp lines starting from 157) implementation does not provide that since if the loading of such certificate fails, the connection is not created.

Additionally the  the peer check seems to be incorrect (Channel.cpp; line 150) since using the value for host verify. A peer is usually the client in the connection and verifying the host/server is not mandatory for  verifying the client. 

A modified (quick edit) source was tested (and examples from rabbitmq-c as well).  Establishing connections works as expected.

**Comment by alanxz** [comment]
I'm willing to accept a pull-request that implements this if you still have the code.

**Comment by mruki** [comment]
Hi,
That should be easy right? Just making a call to "amqp_ssl_socket_set_cacert" optional, based on whether a file is provided.
Tested this on my side and it looks like it works.
@ivaylo-valkov-wrk was it all you've done too?

**Comment by mruki** [comment]
Hi @alanxz,

Eventually, I went with this change to not have to provide CA:

```
diff --git a/src/Channel.cpp b/src/Channel.cpp
index 0d08453..cf7db01 100644
--- a/src/Channel.cpp
+++ b/src/Channel.cpp
@@ -154,16 +154,18 @@ Channel::Channel(const std::string &host, int port, const std::string &username,
 #endif

   try {
-    int status =
-        amqp_ssl_socket_set_cacert(socket, ssl_params.path_to_ca_cert.c_str());
-    if (status) {
-      throw AmqpLibraryException::CreateException(
-          status, "Error setting CA certificate for socket");
+    if (ssl_params.path_to_ca_cert != "") {
+      int status =
+            amqp_ssl_socket_set_cacert(socket, ssl_params.path_to_ca_cert.c_str());
+      if (status) {
+          throw AmqpLibraryException::CreateException(
+              status, "Error setting CA certificate for socket");
+      }
     }

     if (ssl_params.path_to_client_key != "" &&
         ssl_params.path_to_client_cert != "") {
-      status = amqp_ssl_socket_set_key(socket,
+      int status = amqp_ssl_socket_set_key(socket,
                                        ssl_params.path_to_client_cert.c_str(),
                                        ssl_params.path_to_client_key.c_str());
       if (status) {
@@ -172,7 +174,7 @@ Channel::Channel(const std::string &host, int port, const std::string &username,
       }
     }

-    status = amqp_socket_open(socket, host.c_str(), port);
+    int status = amqp_socket_open(socket, host.c_str(), port);
     if (status) {
       throw AmqpLibraryException::CreateException(
           status, "Error setting client certificate for socket");

```

Would be cool if you can review and push to repo if you like it.

Best regards,
Konrad

**Comment by humayunkhan** [comment]
Hi @alanxz,

The solution proposed by @dev-konrad is very useful and eliminates the need for replacing the CA certificate whenever it has expired.

Software that is deployed in organizations where IT support is poor, not having to use the CA certificate will help a lot.

I have used the following code and tested it. It works for me:

int status;
    if (tls_params.ca_cert_path != "") {
      status =
            amqp_ssl_socket_set_cacert(socket, tls_params.ca_cert_path.c_str());
      if (status) {
          throw AmqpLibraryException::CreateException(
              status, "Error setting CA certificate for socket");
      }
     }

I am going to use it (unless you think that something is horribly wrong).

If you feel that this will help a lot of others, consider absorbing this change into your source code.

**Comment by alanxz** [comment]
Resolved by #309
